### PR TITLE
XEH_PreInit (corner cases)

### PR DESCRIPTION
**Pull request by TheMagnetar**
**When merged this pull request will:**
- Move functions defined in main/XEH_preInit.sqf to separate function files.

- [x] EFUNC(lib,getGear) to sys_core/fnc_getGear.sqf. With header.
- [x] EFUNC(lib,replaceGear) to sys_core/fnc_replaceGear.sqf. With header.
- [x] EFUNC(lib,removeGear) to sys_core/fnc_removeGear.sqf. With header.
- [x] EFUNC(lib,addGear) to sys_core/fnc_addGear.sqf. With header.
- [x] EFUNC(lib,getCompartment) to sys_core/fnc_getCompartment.sqf. With header. 
- [x] EFUNC(lib,fastHashCreate) to sys_core/fnc_fastHashCreate.sqf. With header.
- [x] EFUNC(lib,fastHashCopyArray) to sys_core/fnc_fastHashCopyArray.sqf. With header.
- [x] EFUNC(lib,fastHashCopy) to sys_core/fnc_fastHashCopy.sqf. With header.
- [x] EFUNC(lib,fastHashKeys) to sys_core/fnc_fastHashKeys.sqf. With header.

### CODE REVIEW

File: addons/api/fnc_basicMissionSetup.sqf [hunk 1]
```diff
@@ -79,8 +79,8 @@ private _addRadios = {
 
     private _cleanRadioList = [];
     if !("ACRE_PRC343" in _defaultRadios) then {
-        [_player, "ItemRadio"] call EFUNC(lib,removeGear);
-        [_player, "ACRE_PRC343"] call EFUNC(lib,removeGear);
+        [_player, "ItemRadio"] call EFUNC(sys_core, removeGear);
```
Comment (jonpas) [inline addons/api/fnc_basicMissionSetup.sqf]: No space between component name and function name.

File: addons/sys_core/fnc_addGear.sqf [hunk 2]
```diff
@@ -0,0 +1,49 @@
+/*
+ * Author: ACRE2Team
+ * Adds the specified item to a unit. The item can be added to a specific
+ * container (uniform, vest or backpack) or wherever can be added.
+ *
+ * Arguments:
+ * 0: unit <OBJECT>
```
Comment (jonpas) [inline addons/sys_core/fnc_addGear.sqf]: Capitalization, eg. `Unit`, same for the rest.

File: addons/sys_core/fnc_addGear.sqf [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ * 1: item to be added <STRING>
+ * 2: location where the item will be added: vest, uniform or backpack
+ *    <STRING><OPTIONAL>
```
Comment (jonpas) [inline addons/sys_core/fnc_addGear.sqf]: On same line and like that: `... <STRING> (default: "")`

File: addons/sys_core/fnc_addGear.sqf [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+ * [player, "ACRE_PRC343"] call acre_sys_core_fnc_addGear
+ * [player, "ACRE_PRC343", "uniform"] call acre_sys_core_fnc_addGear
+ *
+ * Public: No
+ */
+#include "script_component.hpp"
+
+params ["_unit", "_item",["_gearContainer",""]];
+
+if ( _gearContainer != "") then {
```
Comment (jonpas) [inline addons/sys_core/fnc_addGear.sqf]: Space after opening brace.

File: addons/sys_core/fnc_addGear.sqf [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ * Example:
+ * [player, "ACRE_PRC343"] call acre_sys_core_fnc_addGear
+ * [player, "ACRE_PRC343", "uniform"] call acre_sys_core_fnc_addGear
+ *
+ * Public: No
+ */
+#include "script_component.hpp"
+
+params ["_unit", "_item",["_gearContainer",""]];
```
Comment (jonpas) [inline addons/sys_core/fnc_addGear.sqf]: Missing space after second comma

File: addons/sys_core/fnc_getGear.sqf [hunk 6]
```diff
@@ -0,0 +1,30 @@
+/*
+ * Author: ACRE2Team
+ * Returns an array of ACRE gear in the passed unit
+ *
+ * Arguments:
+ * 0: unit <OBJECT>
+ *
+ * Return Value:
+ * Array of ACRE related gear
```
Comment (jonpas) [inline addons/sys_core/fnc_getGear.sqf]: `ACRE related gear <ARRAY>`

File: addons/sys_core/fnc_replaceGear.sqf [hunk 7]
```diff
@@ -0,0 +1,61 @@
+/*
+ * Author: ACRE2Team
+ * Replaces an item in a unit with another one. It is not limited to ACRE 2
+ * related items
```
Comment (jonpas) [inline addons/sys_core/fnc_replaceGear.sqf]: Same line, this even fits GitHub's width. 😆

File: addons/sys_core/fnc_addGear.sqf [hunk 8]
```diff
@@ -0,0 +1,49 @@
+/*
+ * Author: ACRE2Team
+ * Adds the specified item to a unit. The item can be added to a specific
+ * container (uniform, vest or backpack) or wherever can be added.
```
Comment (jonpas) [inline addons/sys_core/fnc_addGear.sqf]: Entire sentence on second line would look better IMO.

**Review by jonpas — CHANGES_REQUESTED** [review]
Some changes needed.

> Should the Hash functions also be moved to sys_core or Should they remain in main?

Should also be moved.

**Comment by TheMagnetar** [comment]
Missing functions have been added. This is now complete.

File: addons/sys_core/fnc_fastHashCopy.sqf [hunk 9]
```diff
@@ -0,0 +1,39 @@
+/*
+ * Author: ACRE2Team
+ * Copies one hash to another or creates an ACRE2 hash with the object variables if the input is an object
+ *
+ * Arguments:
+ * 0: ACRE 2 Hash <ARRAY> or object <OBJECT>
```
Comment (jonpas) [inline addons/sys_core/fnc_fastHashCopy.sqf]: `* 0: ACRE 2 Hash <ARRAY>/<OBJECT>`

File: addons/sys_core/fnc_getCompartment.sqf [hunk 10]
```diff
@@ -0,0 +1,72 @@
+/*
+ * Author: ACRE2Team
+ * Retrieves the compartment where the specified unit finds itself (driver, turret,...).
+ * Relevant for signal attenuating purposes
```
Comment (jonpas) [inline addons/sys_core/fnc_getCompartment.sqf]: Missing final point.

File: addons/sys_core/fnc_fastHashKeys.sqf [hunk 11]
```diff
@@ -0,0 +1,26 @@
+/*
+ * Author: ACRE2Team
+ * Retrieves a list of keys inside an ACRE 2 hash or variable names defined in an object.
+ * This function can be accessed through the macro HASH_KEYS
```
Comment (jonpas) [inline addons/sys_core/fnc_fastHashKeys.sqf]: Missing final point.

File: addons/sys_core/fnc_fastHashKeys.sqf [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+ *
+ * Arguments:
+ * 0: ACRE 2 hash <HASH> or object <OBJECT>
```
Comment (jonpas) [inline addons/sys_core/fnc_fastHashKeys.sqf]: Probably cleaner that way for future automated docs:
`* 0: ACRE 2 Hash <ARRAY>/<OBJECT>`

File: addons/sys_core/fnc_fastHashCopyArray.sqf [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+ * Copies an array or an ACRE 2 hash entry inside an ACRE 2 hash
+ *
+ * Arguments:
+ * 0: ACRE 2 Hash <HASH> or array <ARRAY> to be copied
+ *
+ * Return Value:
+ * Extracted array of values to be copied <ARRAY>
+ *
+ * Example:
+ * [ARGUMENTS] call acre_sys_core_fnc_fastHashCopyArray
```
Comment (jonpas) [inline addons/sys_core/fnc_fastHashCopyArray.sqf]: Missing `ARGUMENTS` example.

File: addons/sys_core/fnc_fastHashCopy.sqf [hunk 14]
```diff
@@ -0,0 +1,39 @@
+/*
+ * Author: ACRE2Team
+ * Copies one hash to another or creates an ACRE2 hash with the object variables if the input is an object
```
Comment (jonpas) [inline addons/sys_core/fnc_fastHashCopy.sqf]: Missing final point

**Review by jonpas — CHANGES_REQUESTED** [review]
Minor documentation things, good otherwise.

File: addons/sys_core/fnc_fastHashCopyArray.sqf [hunk 15]
```diff
@@ -3,13 +3,14 @@
  * Copies an array or an ACRE 2 hash entry inside an ACRE 2 hash
  *
  * Arguments:
- * 0: ACRE 2 Hash <HASH> or array <ARRAY> to be copied
+ * 0: ACRE 2 Hash <HASH>/<ARRAY> to be copied
```
Comment (jonpas) [inline addons/sys_core/fnc_fastHashCopyArray.sqf]: Vice-versa:
`* 0: ACRE 2 Hash to be copied <HASH>/<ARRAY>`

**Review by jonpas — CHANGES_REQUESTED** [review]

**Review by jonpas — APPROVED** [review]

**Review by Sniperhid — APPROVED** [review]

**Review by Sniperhid — CHANGES_REQUESTED** [review]
It seems when I start the game with this PR my game crashes @jonpas - can you reproduce?

File: addons/main/XEH_preInit.sqf [hunk 16]
```diff
@@ -140,124 +35,20 @@ ACRE_DUMPSTACK_FNC = {
...
-    if (count FAST_HASH_POOL > 0) exitWith {
-        private _ret = (FAST_HASH_POOL deleteAt 0);
-        FAST_HASH_CREATED_HASHES_NEW pushBack _ret;
-        _ret;
-    };
-    private _ret = createLocation ["AcreHashType", [-10000,-10000,-10000], 0, 0];
-    _ret setText "acre_hash";
-    FAST_HASH_CREATED_HASHES_NEW pushBack _ret;
-    _ret;
+EFUNC(sys_core,fastHashCreate) = {
```
Comment (Sniperhid) [inline addons/main/XEH_preInit.sqf]: I'd suggest just adding a comment above this to explain the potential crash if moved.

**Review by Sniperhid — APPROVED** [review]

**Comment by Sniperhid** [comment]
The current gameplan will be to port all sqf code from `main` to `sys_core`. So this is still WIP till it's completed.

File: addons/main/XEH_preInit.sqf [hunk 17]
```diff
@@ -6,8 +6,21 @@ PREP_RECOMPILE_START;
 #include "XEH_PREP.hpp"
 PREP_RECOMPILE_END;
 
-// Lib - @todo put into functions in sys_core and rename to component sys_core across the board
+// Due to dependency with FAST_HASH_CREATED_HASHES_NEW and FAST_HASH_POOL, moving this function to sys_core
+// causes the game to crash during start-up.
+EFUNC(sys_core,fastHashCreate) = {
+  if (count FAST_HASH_POOL > 0) exitWith {
```
Comment (jonpas) [inline addons/main/XEH_preInit.sqf]: 4 spaces indentation.

File: addons/main/fnc__garbageCollector.sqf [hunk 18]
```diff
@@ -0,0 +1,98 @@
+/*
+ * Author: ACRE2Team
+ * ACRE 2 garbage collector. Run as a per frame event handler..
```
Comment (jonpas) [inline addons/main/fnc__garbageCollector.sqf]: One final dot too much.

File: addons/main/fnc__hashMonitor.sqf [hunk 19]
```diff
@@ -0,0 +1,34 @@
+/*
+ * Author: ACRE2Team
+ * Handles deletion and creation of acre hashes.
```
Comment (jonpas) [inline addons/main/fnc__hashMonitor.sqf]: `... ACRE2 hashes.`

File: addons/main/XEH_PREP.hpp [hunk 20]
```diff
@@ -0,0 +1,2 @@
+PREP(_garbageCollector);
```
Comment (jonpas) [inline addons/main/XEH_PREP.hpp]: Let's just make them without leading underscore.

**Review by jonpas — CHANGES_REQUESTED** [review]

**Review by jonpas — APPROVED** [review]

**Pull request merged**